### PR TITLE
GitHub tasks

### PR DESCRIPTION
**Pull request by muhamadazmy**

### CODE REVIEW

File: lib/JumpScale/baselib/atyourservice/ServiceState.py [hunk 1]
```diff
@@ -2,99 +2,109 @@
...
     def methods(self):
         """
         return dict
             key = action name
             val = state
-        state = INIT,ERROR,OK,DISABLED,DO,CHANGED,CHANGEDHRD  DO means: execute the action method as fast as you can, init means it has not been started yet ever
+        state = {state}
+        DO means: execute the action method as fast as you can
+        INIT means it has not been started yet ever
+        """.format(state=','.join(VALID_STATES))
```
Comment (FastGeert) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: Formatting the docstring does not work.

```
In [5]: def test():
   ...:    '''Test %s''' % states
   ...:

In [6]:

In [6]: test?
Signature: test()
Docstring: <no docstring>
File:      c:\users\geert\mobicage\git\appengine\src\<ipython-input-5-a792bd284c2e>
Type:      function

In [7]: def test():
   ...:    '''Test %s'''
   ...:

In [8]:

In [8]: test?
Signature: test()
Docstring: Test %s
File:      c:\users\geert\mobicage\git\appengine\src\<ipython-input-7-7a53468be4f8>
Type:      function

In [9]:
```

File: lib/JumpScale/baselib/atyourservice/ServiceState.py [hunk 2]
```diff
@@ -2,99 +2,109 @@
...
-            self._model["state"][name]=state
-            self.changed=True
-
-    def getSet(self,name,default="DO"):
-        """
-        state = INIT,ERROR,OK,DISABLED,DO,CHANGED,CHANGEDHRD  DO means: execute the action method as fast as you can, init means it has not been started yet ever
+        state = {state}
+        DO means: execute the action method as fast as you can
+        INIT means it has not been started yet ever
+        """.format(state=','.join(VALID_STATES))
```
Comment (FastGeert) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: Formatting docstring does not work

File: lib/JumpScale/baselib/atyourservice/ServiceState.py [hunk 3]
```diff
@@ -2,99 +2,109 @@
...
+        if name not in self._model["state"] or self._model["state"][name] != state:
+            self._model["state"][name] = state
+            self.changed = True
+
+    def getSet(self, name, default="DO"):
         """
+        state = {state}
+        DO means: execute the action method as fast as you can
+        INIT means it has not been started yet ever
+        """.format(state=','.join(VALID_STATES))
```
Comment (FastGeert) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: Formatting docstring does not work

File: lib/JumpScale/baselib/atyourservice/ServiceState.py [hunk 4]
```diff
@@ -148,21 +158,21 @@ def check(self):
         walks over the recurring items and if too old will execute
         """
         from IPython import embed
-        print ("DEBUG NOW check recurring")
+        print("DEBUG NOW check recurring")
```
Comment (FastGeert) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: Replace embed with embed_debugger that does nothing when JS is not run in development mode. Als notify everybody about this (including @despiegk)

File: lib/JumpScale/clients/github/Issue.py [hunk 5]
```diff
@@ -19,6 +20,14 @@ def __init__(self, repo, ddict={}, githubObj=None):
             self.load()
         # self.todo
 
+    @staticmethod
+    def get_story_name(title):
+        m = re_story_name.match(title)
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]: Either strip the title var or make the regex more relax on spaces at the end of the title of an Issue

File: lib/JumpScale/clients/github/Issue.py [hunk 6]
```diff
@@ -71,7 +80,27 @@ def assignee(self):
 
     @property
     def labels(self):
-        return self.ddict["labels"]
+        #we return a copy so changing the list doesn't actually change the ddict value
+        return self.ddict["labels"][:]
+
+    @property
+    def task_estimate(self):
+        m = re_task_estimate.match(self.title)
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]: See title comment. Make the regex more relax on ending spaces or strip the input to the regex.

File: lib/JumpScale/clients/github/Issue.py [hunk 7]
```diff
@@ -71,7 +80,27 @@ def assignee(self):
...
+        if m is not None:
+            return m.group(1)
+        return None
+
+    @property
+    def story_estimate(self):
+        if not len(self.comments):
+            return None, None
+        # find last comment with ETA
+        for i in range(len(self.comments) - 1, -1, -1):
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]: Why not use 
```
for comment in reversed(self.comments):
     blablabla
```

File: lib/JumpScale/clients/github/Issue.py [hunk 8]
```diff
@@ -121,10 +150,10 @@ def type(self):
             if label.startswith("type"):
                 items.append(label)
         if len(items) == 1:
-            return items[0][len("type"):].strip("_")
-        else:
-            self.type = ""
-            return ""
+            _, _, typ = items[0].partition("_")
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]:
```
return items[0].partition('_')[-1]
```

File: lib/JumpScale/clients/github/Issue.py [hunk 9]
```diff
@@ -137,7 +166,8 @@ def priority(self):
             if label.startswith("priority"):
                 items.append(label)
         if len(items) == 1:
-            return items[0][len("priority"):].strip("_")
+            _, _, pri = items[0].partition("_")
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]: See previous comment

File: lib/JumpScale/clients/github/Issue.py [hunk 10]
```diff
@@ -499,14 +537,15 @@ def link_to_story(self, story):
             raise j.exceptions.Input("The task (%s) and the story (%s) have to be in the same Repository." % (self.title, story.task))
             # return
 
-        if not self.isTask:
-            raise j.exceptions.Input("This issue (%s) is not a task" % self.title)
+        # if not self.isTask:
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Issue.py]: Remove commented source code

File: lib/JumpScale/clients/github/Repository.py [hunk 11]
```diff
@@ -51,8 +165,12 @@ def labelnames(self):
 
     @property
     def labels(self):
-        if self._labels is None:
-            self._labels = [item for item in self.api.get_labels()]
+        try:
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: The lock acquire should be outside the try. This is frequently made mistake. If you want to make sure to release something in the finally block, you need to make sure that there is something to release, and you can only do that when you acquire that outside the try finally block.

But you can make it even better by just using the with statement. A nice example can be found here: http://stackoverflow.com/a/5185712

File: lib/JumpScale/clients/github/Repository.py [hunk 12]
```diff
@@ -144,7 +249,7 @@ def labelsSet(self, labels2set,ignoreDelete=["p_"],delete=True):
                         if item.name.startswith(filteritem):
                             ignoreDeleteDo=True
                     if ignoreDeleteDo==False:
-                        from pudb import set_trace; set_trace() 
+                        # from pudb import set_trace; set_trace()
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Remove commented code

File: lib/JumpScale/clients/github/Issue.py (hunk 7 above)
Comment (muhamadazmy) [inline lib/JumpScale/clients/github/Issue.py]: did not know that this exists :+1:

File: lib/JumpScale/clients/github/Repository.py [hunk 13]
```diff
@@ -62,28 +180,15 @@ def stories(self):
         # if not self.fullname.lower().endswith('home'):
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Remove commented source code

File: lib/JumpScale/clients/github/Repository.py [hunk 14]
```diff
@@ -261,24 +364,21 @@ def issues_by_type_state(self, filter=None, collapsepriority=True):
...
+        return GithubRepo.STATES
 
     @property
     def milestones(self):
-        if self._milestones == []:
-            for ms in self.api.get_milestones():
-                milestoneObj = self.getMilestoneFromGithubObj(githubObj=ms)
-                self._milestones.append(milestoneObj)
+        if self._milestones is None:
+            self._milestones = list(map(lambda x: RepoMilestone(self, x), self.api.get_milestones()))
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Map is not very efficient. Consider rewriting as follows:
```
self._milestones = [ RepoMilestone(self, x) for x in self.api.get_milestones() ]
```

File: lib/JumpScale/clients/github/Repository.py [hunk 15]
```diff
@@ -413,6 +498,56 @@ def getColor(self, name):
...
+    def _process_todos(self, issue):
+        priorities_map = {
+            'crit': 'critical',
+            'mino': 'minor',
+            'norm': 'normal',
+            'urge': 'urgent',
+        }
+
+        for todo in issue.todo:
+            try:
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Do not use exceptions like this instead just check if variable todo contains spaces

File: lib/JumpScale/clients/github/Repository.py [hunk 16]
```diff
@@ -413,6 +498,56 @@ def getColor(self, name):
...
+                    self.logger.warning(
+                        'Try to set an non supported priority : %s' % prio)
+                    continue
+
+                prio = "priority_%s" % prio
+                if prio not in issue.labels:
+                    labels = issue.labels
+                    labels.append(prio)
+                    issue.labels = labels
+            # elif cmd == 'delete':
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Remove commented code

File: lib/JumpScale/clients/github/Repository.py [hunk 17]
```diff
@@ -433,12 +568,8 @@ def process_issues(self, issues=[]):
         #     issue.api.create_comment(comment)
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Remove commented code

File: lib/JumpScale/baselib/atyourservice/ServiceState.py (hunk 4 above)
Comment (muhamadazmy) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: @FastGeert what is embed_debugger ? is it something we need to add to jumpscale ? or it exists?

File: lib/JumpScale/clients/github/Repository.py [hunk 18]
```diff
@@ -451,75 +582,153 @@ def get_story(name):
...
+        # generate milestone details page
+        for key, milestone in milestones.items():
+            view = MILESTONE_DETAILS_TEMP.render(repo=self, key=key, milestone=milestone, issues=issues, assignees=assignees, state=state)
+            self.SetFile("milestones/%s.md" % key, view)
+
+        # assignee details page
+        view = ASSIGNEE_REPORT_TMP.render(repo=self, assignees=assignees, state=state)
+        self.SetFile(ASSIGNEE_REPORT_FILE, view)
+
+    def SetFile(self, path, content, message='update file'):
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: Why not set_file ? Are we writing python or java?

File: lib/JumpScale/clients/github/Repository.py [hunk 19]
```diff
@@ -451,75 +582,153 @@ def get_story(name):
...
+        self.api._requester.requestJsonAndCheck(
+            'PUT',
+            self.api.url + '/contents/' + path,
+            input=params,
+        )
+
+    @property
+    def issues(self):
+        try:
+            self._lock.acquire()
```
Comment (FastGeert) [inline lib/JumpScale/clients/github/Repository.py]: See other lock issue

File: lib/JumpScale/clients/github/Issue.py (hunk 6 above)
Comment (muhamadazmy) [inline lib/JumpScale/clients/github/Issue.py]: the regex is already relaxed on ending spaces.

```javascript
/^ETA:\s*(.+)\s*$/
```

File: lib/JumpScale/clients/github/Repository.py (hunk 18 above)
Comment (muhamadazmy) [inline lib/JumpScale/clients/github/Repository.py]: ah, my fault, I was writing to much golang.

File: lib/JumpScale/baselib/atyourservice/ServiceState.py (hunk 4 above)
Comment (FastGeert) [inline lib/JumpScale/baselib/atyourservice/ServiceState.py]: @muhamadazmy @zaibon Lets add embed_debugger into jumscale.
eg
```python
from jumpscale import j

# Start debugging if 1+1 == 2
if 1+1 == 2:
    j.developer.break_into_jshell()
```

**Pull request merged**